src/doc/algorithm/shift_image.cpp

```diff
@@ -20,31 +23,111 @@
 namespace doc {
 namespace algorithm {
 
-void shift_image_with_mask(Image* image, const Mask* mask, int dx, int dy)
+void shift_image(Image* image, int dx, int dy, double angle)
 {
-  gfx::Rect bounds = mask->bounds();
-  if (bounds.isEmpty())
-    return;
-
-  // To simplify the algorithm we use a copy of the original image, we
-  // could avoid this copy swapping rows and columns.
-  ImageRef crop(crop_image(image, bounds.x, bounds.y, bounds.w, bounds.h,
-                           image->maskColor()));
-
-  int u = dx;
-  int v = dy;
-  while (u < 0) u += bounds.w;
-  while (v < 0) v += bounds.h;
-
-  for (int y=0; y<bounds.h; ++y) {
-    for (int x=0; x<bounds.w; ++x) {
-      put_pixel(image,
-                bounds.x + ((u + x) % bounds.w),
-                bounds.y + ((v + y) % bounds.h),
-                get_pixel(crop.get(), x, y));
+  if (cos(angle) < -sqrt(2)/2) {
+    dx = -dx;
+    dy = -dy;
+  }
+  else if (sin(angle) >= sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = -dy;
+    dy = aux;
+  }
+  else if (sin(angle) < -sqrt(2)/2) {
+    double aux;
+    aux = dx;
+    dx = dy;
+    dy = -aux;
+  }
+  Image* imageTmp = Image::create(image->pixelFormat(), image->bounds().w, image->bounds().h);
+  imageTmp->copy(image, gfx::Clip(0, 0, 0, 0, image->bounds().w, image->bounds().h));
```

Comment:
Can we just keep the old code to create the image copy?
```c++
// To simplify the algorithm we use a copy of the original image, we
// could avoid this copy swapping rows and columns.
ImageRef crop(crop_image(image, bounds.x, bounds.y, bounds.w, bounds.h,
                         image->maskColor()));	
```
Also there is a memory leak here because the `imageTmp` is a raw pointer (`Image*`, which requires a manual `delete` which is missing). Use a smart pointer in these cases (`ImageRef`) like in the old code (I would use most of the old code as possible).